nomad/resource_quota_specification.go

```diff
@@ -0,0 +1,270 @@
+
+	spec := api.QuotaSpec{
+		Name:        d.Get("name").(string),
+		Description: d.Get("description").(string),
+	}
+	limits, err := expandQuotaLimits(d)
+	if err != nil {
+		return err
+	}
+	spec.Limits = limits
+
+	// upsert our quota spec
+	log.Printf("[DEBUG] Creating quota specification %q", spec.Name)
+	_, err = client.Quotas().Register(&spec, nil)
+	if err != nil {
+		return fmt.Errorf("error registering quota specification %q: %s", spec.Name, err.Error())
+	}
+	log.Printf("[DEBUG] Created quota specification %q", spec.Name)
+	d.SetId(spec.Name)
+
+	return resourceQuotaSpecificationRead(d, meta)
+}
+
+func resourceQuotaSpecificationUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	spec := api.QuotaSpec{
+		Name:        d.Get("name").(string),
+		Description: d.Get("description").(string),
+	}
+	limits, err := expandQuotaLimits(d)
+	if err != nil {
+		return err
+	}
+	spec.Limits = limits
+
+	// upsert our quota spec
+	log.Printf("[DEBUG] Updating quota specification %q", spec.Name)
+	_, err = client.Quotas().Register(&spec, nil)
+	if err != nil {
+		return fmt.Errorf("error registering quota specification %q: %s", spec.Name, err.Error())
+	}
+	log.Printf("[DEBUG] Updated quota specification %q", spec.Name)
+
+	return resourceQuotaSpecificationRead(d, meta)
+}
+
+func resourceQuotaSpecificationDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	name := d.Id()
+
+	// delete the quota spec
+	log.Printf("[DEBUG] Deleting quota specification %q", name)
+	_, err := client.Quotas().Delete(name, nil)
+	if err != nil {
+		return fmt.Errorf("error deleting quota specification %q: %s", name, err.Error())
+	}
+	log.Printf("[DEBUG] Deleted quota specification %q", name)
+
+	return nil
+}
+
+func resourceQuotaSpecificationRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	name := d.Id()
+
+	// retrieve the policy
+	log.Printf("[DEBUG] Reading quota specification %q", name)
+	spec, _, err := client.Quotas().Info(name, nil)
+	if err != nil {
+		// we have Exists, so no need to handle 404
+		return fmt.Errorf("error reading quota specification %q: %s", name, err.Error())
+	}
+	log.Printf("[DEBUG] Read quota specification %q", name)
+
+	d.Set("name", spec.Name)
+	d.Set("description", spec.Description)
+	err = d.Set("limits", flattenQuotaLimits(spec.Limits))
+	if err != nil {
+		return fmt.Errorf("error setting quota specification limits for %q: %s", name, err.Error())
+	}
+
+	return nil
+}
+
+func resourceQuotaSpecificationExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	client := meta.(*api.Client)
+
+	name := d.Id()
+	log.Printf("[DEBUG] Checking if quota specification %q exists", name)
+	resp, _, err := client.Quotas().Info(name, nil)
+	if err != nil {
+		// As of Nomad 0.4.1, the API client returns an error for 404
+		// rather than a nil result, so we must check this way.
+		if strings.Contains(err.Error(), "404") {
```

Comment:
This is super ugly :cry:  is there an issue open on the Nomad API to at least return a named/typed err?

Reply:
There is! https://github.com/hashicorp/nomad/issues/1849